src/game.zig

```diff
@@ -277,6 +278,43 @@ pub const collision = struct {
 		return resultBox;
 	}
 
+	pub fn calculateFriction(comptime side: main.utils.Side, pos: Vec3d, hitBox: Box, defaultFriction: f32) f32 {
+		const boundingBox: Box = .{
+			.min = pos + hitBox.min,
+			.max = pos + hitBox.max,
+		};
+		const minX: i32 = @intFromFloat(@floor(boundingBox.min[0]));
+		const maxX: i32 = @intFromFloat(@floor(boundingBox.max[0] - 0.0001));
+		const minY: i32 = @intFromFloat(@floor(boundingBox.min[1]));
+		const maxY: i32 = @intFromFloat(@floor(boundingBox.max[1] - 0.0001));
+
+		const areaBottom = (boundingBox.max[0] - boundingBox.min[0]) * (boundingBox.max[1] - boundingBox.min[1]);
+		
+		const z: i32 = @intFromFloat(@floor(boundingBox.min[2] - 0.01));
+
+		var friction: f32 = 0;
+
+		var x: i32 = minX;
+		while (x <= maxX) : (x += 1) {
+			var y: i32 = minY;
+			while (y <= maxY) : (y += 1) {
+				const _block = if(side == .client) main.renderer.mesh_storage.getBlock(x, y, z)
+				else main.server.world.?.getBlock(x, y, z);
+				const min = @max(Vec2d{@floatFromInt(x), @floatFromInt(y)}, vec.xy(boundingBox.min));
+				const max = @min(Vec2d{@floatFromInt(x + 1), @floatFromInt(y + 1)}, vec.xy(boundingBox.max));
+				const area = (max[0] - min[0]) * (max[1] - min[1]);
+				
+				if (_block) |block| {
+					friction += @as(f32, @floatCast(area / areaBottom)) * (1 / block.friction());
```

Comment:
Friction should be averaged over just the contact area, not the total area.
If you stand on a ledge of a block it should give the same friction as if you stand on a full block.